### PR TITLE
Catch ParentMissingError in variables checker

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
add compatibility with https://github.com/pylint-dev/astroid/pull/2345

some of the unrelated changes here are just for the sake of passing pylint too-many-statements, etc

### CODE REVIEW

**Review by nickdrozd — APPROVED** [review]
Fixes the problem for me!

**Comment by jacobtylerwalls** [comment]
I think once we bump astroid these lines will be covered. I'm happy to wait to merge until then.

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -2547,8 +2547,12 @@ def _loopvar_name(self, node: astroid.Name) -> None:
         else:
             _astmts = astmts[:1]
         for i, stmt in enumerate(astmts[1:]):
-            if astmts[i].statement().parent_of(stmt) and not utils.in_for_else_branch(
-                astmts[i].statement(), stmt
+            try:
+                astmt_statement = astmts[i].statement()
+            except astroid.exceptions.ParentMissingError:
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2552]: We should add `import astroid.exceptions` to the top of this file.

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -2596,23 +2600,24 @@ def _loopvar_name(self, node: astroid.Name) -> None:
...
+                comprehension, nodes.ComprehensionScope
+            )
+            if comprehension_scope is None:
+                # Should not be possible.
+                pass
+            elif (
+                comprehension_scope.parent.scope() is scope
+                and node.name in comprehension_scope.locals
+            ):
+                return
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2620]: Tbh I'm not sure this is really an improvement if it's just to work around the `too-many-statements` error. We already disable `too-many-branches` for this function, why not `too-many-statements` as well.

https://github.com/pylint-dev/pylint/blob/51c28bc922dfb8da42e680ed7190eb942889df01/pylint/checkers/variables.py#L2524-L2525

```diff
-    # pylint: disable = too-many-branches
+    # pylint: disable-next=too-many-branches,too-many-statements
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
Encountered the same error @nickdrozd has while testing. I'd recommend we merge this once all comments are addressed. The `astroid` update wouldn't be able to be merged before that anyway as with it some test cases would be failing (which are fixed with this change).

File: pylint/checkers/variables.py [hunk 3]
```diff
@@ -2649,8 +2654,7 @@ def _loopvar_name(self, node: astroid.Name) -> None:
                 self.add_message("undefined-loop-variable", args=node.name, node=node)
                 return
 
-            elements = getattr(inferred, "elts", getattr(inferred, "items", []))
-            if not elements:
+            if not getattr(inferred, "elts", getattr(inferred, "items", [])):
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2657]: Same here. Add the disable and revert afterwards.

File: pylint/checkers/variables.py [hunk 4]
```diff
@@ -2511,7 +2512,7 @@ class D(Tp):
             and name in frame_locals
         )
 
-    # pylint: disable = too-many-branches
+    # pylint: disable = too-many-branches, too-many-statements
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2515]:
```suggestion
    # pylint: disable-next=too-many-branches,too-many-statements
```
The whitespace is up to you. I would just recommend to use `disable-next`.

**Review by cdce8p — APPROVED** [review]
Looks good! Thanks @jacobtylerwalls 👍🏻

**Pull request merged**